layer4/listener.go

```diff
@@ -114,25 +116,35 @@ type listener struct {
 	logger        *zap.Logger
 	compiledRoute Handler
 
+	closed atomic.Bool
```

Comment:
Would a [sync.Once](https://pkg.go.dev/sync#Once) be simpler here?

Reply:
`sync.Once` is used to make sure something happens once, not used to test if something happened. It doesn't fit in the struct

Reply:
Ah, I see below, the done channel is used for signaling.